keps/sig-node/20190322-image-gc-whitelist.md

```diff
@@ -0,0 +1,273 @@
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Examples](#examples)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+      - [Removing a deprecated flag](#removing-a-deprecated-flag)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Enhancing Kubelet image garbage collector with a whitelist of container images. 
+Whitelisted images will be ignored during garbage collection.
+Currently only the dockershim pod sandbox image is whitelisted, this KEP aims
+to make the whitelist configurable.
+
+## Motivation
+
+There are some use cases when not all the container images are available from a remote repository, therefore container images are stored locally in the image cache of the Nodes. These container images can be preloaded to the Nodes instead of pulled on demand.
+The image garbage collector in Kubelet deletes all unused images from the image cache in case of storage shortage.
+When a Pod using such preloaded image scheduled to a Node from where the container image was garbage collected the Pod will go to ImagePullBackOff state forever.
+
+Users of container runtimes other than dockershim may wish to whitelist the
+podsandbox image used by their container runtime, which may not be in sync
+with dockershim.
+
+### Goals
+
+- Add a Kubelet command line option to list whitelisted container images
+- Skip the garbage collection of the whitelisted container images
+
+### Non-Goals
+
+- Further complicating the kubelet image GC algorithm.
+  - Kubelet already has this logic for the dockershim pause image, we just need
+  to apply it to a configurable list. We do not aim to add any further complexity.
+
+## Proposal
+
+### User Stories
+
+The user stories in the following chapters are alternatives of each other.
+
+#### Story 1
```

Comment:
Per the original KEP discussion these user stories in particular need fleshing out.

Reply:
- cluster maintenance cronjob that needs to run on every node, and therefore looks unused to the gc most of the time.
- my container registry is broken and certain images need to be obtained out of band
- I run Docker-era legacy workloads that build a host-customized container image, and I need the Kubernetes machinery to not delete the image during the brief periods when the Pod is being recreated and appears unused.